solarextrapolation/potential_extrapolator.py

```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jun 12 13:01:54 2015
+
+@author: Alex
+"""
+# Import the SolarBExtrapolation API
+from classes import *
+import numpy as np
+
+class PotentialExtrapolator(Extrapolators):
+    def __init__(self, map_magnetogram, **kwargs):
+        super(PotentialExtrapolator, self).__init__(map_magnetogram, **kwargs)
+        self.meta['extrapolator_routine'] = 'Potential Field Extrapolator'
+
+    def _extrapolation(self):
+        # Adding in custom parameters to the meta
+        phi = self._extrapolate_phi()
+        Bxyz = self._determine_vec(phi, D = 1, debug = False)
+
+        return Map3D( Bxyz, self.meta )
+
+    # Greens function.
+    def _Gn_5_2_26(self, inR, inRPrime):    
+        floModDr = np.linalg.norm(inR - inRPrime)
+        floOut = 1.0 / ( 2.0 * np.pi * floModDr)
+        return floOut
+    
+    # Greens function.
+    def _Gn_5_2_29(self, i, j, k, i_prime, j_prime, d, d_com):    
+        d_i = i - i_prime
+        d_j = j - j_prime
+        d_k = k - d_com # d / np.sqrt(2.0 * np.pi)
+        floModDr = np.sqrt( d_i * d_i + d_j * d_j + d_k * d_k)
+        
+        #print 'floModDr: ' + str(floModDr)
+        floOut = 1.0 / ( 2.0 * np.pi * floModDr)
+        #print 'floOut: ' + str(floOut)
+        return floOut
+    
+    
+    # A function to extrapolate the magnetic field above the given boundary.
+    def _phi_extrapolation_python(self, boundary, d):    
+        # Volume size.
+        M = boundary.shape[0]
+        N = boundary.shape[1]
+        Z = self.z
+        
+        # Derived parameters
+        d = 1.0
+        d_squ = np.power(d,2)
+        d_com = d / np.sqrt(2.0 * np.pi)
+        
+        # Create the empty numpy volume array.
+        D = np.empty((M, N, Z), dtype=np.float)
+        
+        # Itherate though the 3D space.
+        for i in range(0, M):
+            for j in range(0, N):
+                for k in range(0, Z):
+                    # Variable holding running total for the contributions to point.
+                    point_phi_sum = 0.0
+                    # Iterate through the boundary data.
+                    for i_prime in range(0,M):
+                        for j_prime in range(0,N):                        
+                            # Find the components for this contribution product.
+                            B_n = boundary[i_prime, j_prime]
+                            
+                            G_n = self._Gn_5_2_29(i, j, k, i_prime, j_prime, d, d_com)
```

Comment:
I wonder if you could do this using numpy ufuncs http://docs.scipy.org/doc/numpy/reference/ufuncs.html

It looks like you could, which would then vectorise it for you.